File: binding-python/runtime/src/main/python/etch/binding/msg/Type.py
"""
#  Licensed to the Apache Software Foundation (ASF) under one   *
#  or more contributor license agreements.  See the NOTICE file *
#  distributed with this work for additional information        *
#  regarding copyright ownership.  The ASF licenses this file   *
#  to you under the Apache License, Version 2.0 (the            *
#  "License"); you may not use this file except in compliance   *
#  with the License.  You may obtain a copy of the License at   *
#                                                               *
#    http://www.apache.org/licenses/LICENSE-2.0                 *
#                                                               *
#  Unless required by applicable law or agreed to in writing,   *
#  software distributed under the License is distributed on an  *
#  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
#  KIND, either express or implied.  See the License for the    *
#  specific language governing permissions and limitations      *
#  under the License. 
"""
from __future__ import absolute_import

import types
from .IdName                  import *
from ...python.Exceptions     import *
from .FieldMap       import *
from .ComboValidator import *
from .AsyncMode      import *
from .Direction      import *

class Type(IdName):
    """
    Type denotes the type of a struct or message. When used with a message it
    typically denotes an action or event.
    """

    def __init__(self, *args, **kwargs):
        """
        Constructs the type

        @param type_id    The id of the type
        @param type_name  The name of the type
        """
        super(Type,self).__init__(*args, **kwargs)

        self.__rType           = None
        self.__helper          = None
        self.__clss            = None
        self.__locked          = False
        self.__timeout         = 0
        self.__responseField   = None
        self.__stubHelper      = None
        self.__superType       = None
        self.__asyncMode       = AsyncMode.NONE
        self.__direction       = Direction.BOTH
        self.__validators      = {}    # TODO: self.__validators should be synchronized?
        self.__fields          = FieldMap()

    def getValidator(self, key):
        """
        @param key
        @return       The validator for this key of this type
        """
        return self.__validators.get(key, None)

    def putValidator(self, key, vldtr):
        """
        Adds the validator to the chain for this key

        @param    key
        @param    vldtr
        """
        self.checkNotLocked()

        if vldtr == None:
            return;

        if self.__fields.get(key.getId()) == None:
            self.addField(key)

        v = self.__validators.get(key, None)
        if v != None:
            self.__validators[key] = ComboValidator(v, vldtr)
        else:
            self.__validators[key] = vldtr

    def clearValidator(self, key):
        """
        Removes the validator chain for this key

        @param key
        """
        self.checkNotLocked()
        del self.__validators[key]

    def getResult(self):
        """
        @return   The result message type of this message type
        """
        return self.__rType

    def setResult(self, rType):
        """
        Sets the result message type of this message type

        @param rType
        """
        self.checkNotLocked()
        self.__rType = rType

    def getImportExportHelper(self):
        """
        @return       The associated import/export helper
        """
        return self.__helper

    def setImportExportHelper(self, helper):
        """
        Sets the associated import/export helper

        @param helper
        """
        self.checkNotLocked()
        self.__helper = helper


    def getComponentType(self):
        """
        @return the associated component type for an array of this type
        """
        return self.__clss

    def setComponentType(self, clss):
        """
        Sets the associated component type fro an array of this type.

        This type is not used when de-serializing the array component, just
        when allocating the array itself.

        @param clss
        """
        self.checkNotLocked()
        self.__clss = clss


    def getStubHelper(self):
        """
        @return       An object to help the stub dispatch the received message
        """
        return self.__stubHelper

    def setStubHelper(self, stubHelper):
        """
        Sets an object to help the stub dispatch the recieved message

        @param stubHelper
        """
        if self.__stubHelper != None:
            raise IllegalStateException, "stubHelper is already set"
        self.__stubHelper = stubHelper

    def lock(self):
        """
        Locks the fields for this type
        """
        self.__locked = True
        self.__fields.lock()

    def checkNotLocked(self):
        if self.__locked:
            raise IllegalStateException, "locked"

    def addField(self, field):
        """
        Adds a field to the set of fields

        @param field      A Field to add
        @return           The argument
        @throws           IllegalArguementException
        """
        return self.__fields.add(field)

    def getField(self, key):
        """
        Translates a field id or name into the appropriate Field object

        @param key    A Field id or name
        @return       Id translated into the appropriate Field
        """
        return self.__fields.get(key)

    def getFields(self):
        """
        @return       A set of all the fields
        """
        return set(self.__fields.values())

    def getTimeout(self):
        """
        @return     The time in milliseconds to wait for this response message
        """
        return self.__timeout
    
    def setTimeout(self, timeout):
        """
        @param timeout  the time in milliseconds to wait for this response message
        """
        self.checkNotLocked()
        if not isinstance(timeout, (types.LongType, types.IntType)):
            raise IllegalArgumentException, "timeout must be an integer"
        if timeout < 0:
            raise IllegalArgumentException, "timeout cannot be negative"
            
        self.__timeout = timeout
        
    def getResponseField(self):
        """
        @return the field with the value of this response.
        """
        return self.__responseField

    def setResponseField(self, responseField):
        """
        @param responseField    the field with the value of this response
        """
        self.checkNotLocked()
        self.__responseField = responseField
    
    def isAssignableFrom(self, other):
        """
        Checks whether this type is assignment compatible with other. This 
        means that the other is a subclass of this one

        @param other
        @return     True if this type is assignable from other
        """
        return other != None and (self == other or self.isAssignableFrom(other.superType()))

    def checkIsAssignableFrom(self, other):
        """
        Checks to see if this type is assignable from the other type.

        @param other
        @raises IllegalArgumentException if self is not assignable from otherType

        """
        if not self.isAssignableFrom(other):
            raise IllegalArgumentException, "This is not a super type of other"
            
    def superType(self):
        """
        @return the super type of this type. If struct A extends B, then B is the super type of A
        """
        return self.__supertype

    def setSuperType(self, superType):
        """
        Sets the super type of this type. If struct A extends B, then B is the super type of A.
        @param superType
        """
        self.checkNotLocked()
        self.__superType = superType
    
    def getAsyncMode(self):
        """
        @return the AsyncMode for this type.
        """
        return self.__asyncMode
  
    def setAsyncMode(self, mode):
        """
        Sets the AsyncMode for this type.
        @param mode
        """
        self.checkNotLocked()
        self.__asyncMode = mode
    
    def getDirection(self):
        """
        @return the message direction
        """
        return self.__direction
    
    def setDirection(self, direction):
        """
        Sets the message direction
        @param direction the direction to set
        """
        if self.checkNotLocked():
            self.__direction = direction



    

File: binding-java/runtime/src/main/java/org/apache/etch/bindings/java/msg/Type.java
/* $Id$
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.etch.bindings.java.msg;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;

/**
 * Type denotes the type of a struct or message. When used with a
 * message it typically denotes an action or event.
 */
public final class Type extends IdName
{
	/**
	 * Constructs the Type.
	 * @param id the id of the type.
	 * @param name the name of the type.
	 */
	public Type( int id, String name )
	{
		super( id, name );
	}

	/**
	 * Constructs the Type, computing the appropriate value
	 * for the id.
	 * @param name the name of the type.
	 */
	public Type( String name )
	{
		super( name );
	}

	/**
	 * @param key
	 * @return the validator for this key of this type.
	 */
	public Validator getValidator( Field key )
    {
	    return validators.get( key );
    }
	
	/**
	 * Adds the validator to the chain for this key.
	 * @param key
	 * @param vldtr
	 */
	public void putValidator( Field key, Validator vldtr )
	{
		checkNotLocked();
		
		if (vldtr == null)
			return;
		
		if (fields.get( key.getId() ) == null)
			addField( key );

		Validator v = validators.get( key );
		if (v != null)
			validators.put( key, new ComboValidator( v, vldtr ) );
		else
			validators.put( key, vldtr );
	}

	/**
	 * Removes the validator chain for this key.
	 * @param key
	 */
	public void clearValidator( Field key )
    {
		checkNotLocked();
	    validators.remove( key );
    }
	
	private final Map<Field, Validator> validators = new HashMap<Field, Validator>();
	
	/**
	 * @return the result message type of this message type.
	 */
	public Type getResult()
	{
		return rType;
	}

	/**
	 * Sets the result message type of this message type.
	 * @param rType
	 */
	public void setResult( Type rType )
	{
		checkNotLocked();
		this.rType = rType;
	}
	
	private Type rType;

	/**
	 * @return the associated import / export helper.
	 */
	public ImportExportHelper getImportExportHelper()
	{
		return helper;
	}
	
	/**
	 * Sets the associated import / export helper.
	 * @param helper
	 */
	public void setImportExportHelper( ImportExportHelper helper )
	{
		checkNotLocked();
		this.helper = helper;
	}
	
	private ImportExportHelper helper;

	/**
	 * @return the associated component type for an array of this type.
	 * This type is not used when de-serializing the array components,
	 * just when allocating the array itself.
	 */
	public Class<?> getComponentType()
	{
		return clss;
	}
	
	/**
	 * Sets the associated component type for an array of this type.
	 * This type is not used when de-serializing the array components,
	 * just when allocating the array itself.
	 * @param clss
	 */
	public void setComponentType( Class<?> clss )
	{
		checkNotLocked();
		this.clss = clss;
	}
	
	private Class<?> clss;

	/**
	 * @return an object to help the stub dispatch the received message.
	 */
	public Object getStubHelper()
	{
		return stubHelper;
	}
	
	/**
	 * Sets an object to help the stub dispatch the received message.
	 * @param stubHelper
	 */
	public void setStubHelper( Object stubHelper )
	{
//		checkNotLocked(); called from stub and not from value factory.
		if (this.stubHelper != null)
			throw new IllegalStateException( "this.stubHelper != null" );
		this.stubHelper = stubHelper;
	}

	private Object stubHelper;
	
	/**
	 * Locks the fields for this type.
	 */
	public void lock()
	{
		locked = true;
		fields.lock();
	}
	
	private void checkNotLocked()
	{
		if (locked)
			throw new IllegalStateException( "locked" );
	}
	
	private boolean locked;
	
	/**
	 * Adds a field to the set of fields.
	 * 
	 * @param field a field to add.
	 * 
	 * @return the argument. If there is a collision with
	 * an id and name, both associated with the same field,
	 * then that field is returned instead of the argument.
	 * 
	 * @throws IllegalArgumentException if there is a
	 * collision in the id or name, or both id and name
	 * when not associated with the same field.
	 */
	public Field addField( Field field )
	{
		return fields.add( field );
	}

	/**
	 * Translates a field id into the appropriate Field object.
	 * @param id a field id.
	 * @return id translated into the appropriate Field.
	 */
	public Field getField( Integer id )
	{
		return fields.get( id );
	}

	/**
	 * Translates a field name into the appropriate Field
	 * object.
	 * @param name a field name.
	 * @return name translated into the appropriate Field.
	 */
	public Field getField( String name )
	{
		return fields.get( name );
	}
	
	/**
	 * @return a set of all the fields.
	 */
	public Set<Field> getFields()
	{
		return Collections.unmodifiableSet( fields.values() );
	}
	
	private final FieldMap fields = new FieldMap();

	/**
	 * @return the time in milliseconds to wait for this response message.
	 */
	public int getTimeout()
	{
		return timeout;
	}

	/**
	 * @param timeout the time in milliseconds to wait for this response message.
	 */
	public void setTimeout( int timeout )
	{
		checkNotLocked();
		this.timeout = timeout;
	}
	
	private int timeout;

	/**
	 * @return the field with the value of this response.
	 */
	public Field getResponseField()
	{
		return responseField;
	}

	/**
	 * @param responseField the field with the value of this response.
	 */
	public void setResponseField( Field responseField )
	{
		checkNotLocked();
		this.responseField = responseField;
	}
	
	private Field responseField;
	
	/**
	 * Checks whether this type is assignment compatible with other. This
	 * means that other is a subclass of this.
	 * @param other
	 * @return true if this type is assignable from other.
	 */
	public boolean isAssignableFrom( Type other )
	{
		return other != null && (equals( other ) || isAssignableFrom( other.superType() ));
	}

	/**
	 * Checks to see if this type is assignable from the other type.
	 * @param other
	 * @throws IllegalArgumentException if this not assignable from other.
	 * @see #isAssignableFrom(Type)
	 */
	public void checkIsAssignableFrom( Type other )
	{
		if (!isAssignableFrom( other ))
			throw new IllegalArgumentException( "this not super type of other" );
	}

	/**
	 * @return the super type of this type. If struct A extends B, then
	 * B is the super type of A.
	 */
	public Type superType()
	{
		return superType;
	}
	
	/**
	 * Sets the super type of this type. If struct A extends B, then
	 * B is the super type of A.
	 * @param superType
	 */
	public void setSuperType( Type superType )
	{
		checkNotLocked();
		this.superType = superType;
	}
	
	private Type superType;

	/**
	 * @return the AsyncMode for this type.
	 */
	public AsyncMode getAsyncMode()
	{
		return asyncMode;
	}
	
	/**
	 * Sets the AsyncMode for this type.
	 * @param mode
	 */
	public void setAsyncMode( AsyncMode mode )
	{
		checkNotLocked();
		asyncMode = mode;
	}
	
	private AsyncMode asyncMode;

	/**
	 * @return the message direction.
	 */
	public Direction getDirection()
	{
		return direction;
	}

	/**
	 * Sets the message direction.
	 * @param direction the direction to set.
	 */
	public void setDirection( Direction direction )
	{
		checkNotLocked();
		this.direction = direction;
	}
	
	private Direction direction;
}
